Add new tail sampling processor policy: status_code

yvrhdn opened the pull request:
**Description:** 
This adds a new policy to the tail sampling processor sampling traces based upon the status code of a span. Adding this policy allows you to sample all traces of which the status code matches a given status code.

~~**Link to tracking Issue:** <Issue number if applicable>~~

**Testing:** <Describe what testing was performed and which tests were added.>

Added unit tests similar to the other policies.

~~⚠️  I haven't been able to write proper tests because I can't figure out how to set the status code on spans manually. All the `pdata` data structures are closed off: fields are private and there is not setter for status code 🤷🏻  If there is a mechanism to still set it, please let me know and I'll update the tests!~~

**Documentation:**

I have updated the README.

Commit:
tailsamplingprocessor: add status code policy

kvrhdn commented:
We won't be able to merge this PR and #3750 at the same time because of merge conflicts, but I can resolve any conflicts in whatever PR we merge second.

jpkrohling commented on processor/tailsamplingprocessor/sampling/status_code.go:
```diff
@@ -0,0 +1,102 @@
+					span := ils.Spans().At(j)
+
+					for _, statusCode := range r.statusCodes {
+						if span.Status().Code() == statusCode {
```
Don't we have status codes as attributes in the resource span attributes?

Also, this code seems now repeated across policies. What do you think of sending a small PR to be merged before this one, extracting most of this logic into its own function? The new function can accept a `rspans` and a function that returns whether the condition was found. Like (pseudo code):

```go
func hasSpanWithCondition(rspans pdata.ResourceSpanSlice, func(pdata.Span) bool) bool {}

for _, batch := range batches {
  rspans := batch.ResourceSpans()
  if hasSpanWithCondition(rspans, func(span pdata.Span) bool {
    for _, statusCode := range r.statusCodes {
      if span.Status().Code() == statusCode {
        return true
      }
    }
    return false
  }) {
    return Sampled, nil
  }
}
```

kvrhdn replied:
> Don't we have status codes as attributes in the resource span attributes?

I'm not sure, I'll have to familiarise myself a bit more with the difference between the resource span attributes and the instrumentation library spans.

> Also, this code seems now repeated across policies. What do you think of sending a small PR to be merged before this one, extracting most of this logic into its own function? The new function can accept a rspans and a function that returns whether the condition was found. Like (pseudo code):

Agree :+1: Are you planning to merge #3750? I can make a PR incorporating those changes after it has been merged.

jpkrohling replied:
I'm just an approver, we need someone with super powers: @tigrannajaryan, for instance :-)

kvrhdn replied:
Avoiding repeated code has been addressed in #3787

jpkrohling reviewed (commented):
>  I haven't been able to write proper tests because I can't figure out how to set the status code on spans manually. All the pdata data structures are closed off: fields are private and there is not setter for status code

Doesn't this one here work?

https://github.com/open-telemetry/opentelemetry-collector/blob/0d747472e565e6548ebfe42a3b3205d29d9e9065/consumer/pdata/traces.go#L150-L162

kvrhdn commented:
> Doesn't this one here work?
> 
> https://github.com/open-telemetry/opentelemetry-collector/blob/0d747472e565e6548ebfe42a3b3205d29d9e9065/consumer/pdata/traces.go#L150-L162

🤯  yeah that works as you'd expect, not sure how I missed it 😅

Commit:
Improve tests status_code tail sampling policy

Commit:
Refactor tail sampling processor policies

Move repeated boiler plate logic related to iterating through traces into a common set of functions.

Commit:
Add new tail sampling processor latency to changelog

Commit:
Add license

kvrhdn commented:
Rebased on top of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3787

Commit:
Refactor using util helper function

kvrhdn commented:
I'm going to pull this comment out of the review so it's easier to continue the discussion:

> Don't we have status codes as attributes in the resource span attributes?

I'm not sure, I can't find anything that would say so. According to the spec the status is a separate field on a span ([opentelemetry-specification/specification/trace/api.md#span](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span)).

I've been digging through the various resource attribtues, but haven't found anything related to the span status or status code. ([opentelemetry-specification/specification/resource/semantic_conventions/README.md](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/README.md))

Commit:
Add new tail sampling processor status_code to changelog

kvrhdn commented:
(I'm going to rebase this again after #3787 and #3788 get merged, this will clean up the commits)

kvrhdn commented:
Friendly bump, this PR is still active I'm just waiting for a review 🙂

mx-psi commented:
@kvrhdn can you fix the merge conflict?

@jpkrohling can you have another look?

Commit:
Merge branch 'main' into status_code

# Conflicts:
#	CHANGELOG.md

jpkrohling commented on CHANGELOG.md:
```diff
@@ -4,7 +4,9 @@
 
-- `tailsampling` processor: Add new policy `latency` (#3750)
+- `tailsampling` processor:
+  - Add new policy `latency` (#3750)
```
I guess the latency made it to 0.29.0, no?

jpkrohling replied:
Actually... 0.29.0 hasn't been released for contrib yet. I heard yesterday that the release train has departed already, though.

kvrhdn replied:
Is it still possible to get this PR in 0.29 or is the release already cut?

kvrhdn replied:
I'm not sure how to avoid a merge conflict with the release PR, should I wait until #3863 is merged to rebase and update the changelog?

jpkrohling replied:
I think @bogdandrutu mentioned yesterday that they were having build problems, blocking the release. I don't expect any new features to be included for 0.29.0. Wait for the release, then rebase and update the changelog.

jpkrohling commented on processor/tailsamplingprocessor/sampling/status_code.go:
```diff
@@ -0,0 +1,85 @@
+		case "UNSET":
+			statusCodes[i] = pdata.StatusCodeUnset
+		default:
+			return nil, fmt.Errorf("unknown status code %s, supported: OK, ERROR, UNSET", statusCodeString)
```
```suggestion
			return nil, fmt.Errorf("unknown status code %q, supported: OK, ERROR, UNSET", statusCodeString)
```

kvrhdn replied:
Just noticed there was a bug in the error handling, fixed this now.

jpkrohling reviewed (commented):
Looks good, the changelog needs to be changed though.

Commit:
Update processor/tailsamplingprocessor/sampling/status_code.go

Co-authored-by: Juraci Paixão Kröhling <juraci.github@kroehling.de>

Commit:
Fix error handling
